sip/sip-002-smart-contract-language.md

```diff
@@ -0,0 +1,600 @@
+transaction on the blockchain which invokes the public function.
+
+This smart contracting language differs from most other smart
+contracting languages in two important ways:
+
+1. The language _is not_ intended to be compiled. The LISP language
+   described in this document is the specification for correctness.
+2. The language _is not_ Turing complete. This allows us to guarantee
+   that static analysis of programs to determine properties like
+   runtime cost and data usage can complete successfully.
+
+## Specifying Contracts
+
+A smart contract definition is specified in a LISP language with the
+following limitations:
+
+1. Recursion is illegal and there is no `lambda` function.
+2. Looping may only be performed via `map`, `filter`, or `fold`
+3. The only atomic types are booleans, integers, fixed length
+   buffers, and principals
+4. There is additional support for lists of the atomic types, however
+   the only variable length lists in the language appear as function
+   inputs (i.e., there is no support for list operations like append
+   or join).
+5. Variables may only be created via `let` binding and there
+   is no support for mutating functions like `set`.
+6. Defining of constants and functions are allowed for simplifying
+   code using `define` statement. However, these are purely
+   syntactic. If a definition cannot be inlined, the contract will be
+   rejected as illegal. These definitions are also _private_, in that
+   functions defined this way may only be called by other functions
+   defined in the given smart contract.
+7. Functions specified via `define-public` statements are _public_
+   functions. Arguments to these functions must specify their types.
+
+Public functions return a boolean result. If the function returns
+`true`, then it is considered valid, and any changes made to the
+blockchain state will be materialized. If the function returns
+`false`, it will be considered invalid, and will have _no effect_ on
+the smart contract's state, except for a transaction fee debit.
+
+## Inter-Contract Calls
+
+A smart contract may call functions from other smart contracts using a
+`(contract-call)` function. This function accepts a function name and
+the smart contract's _identifier_ as input.  A smart contract's
+identifier is a hash of the smart contract's definition, represented
+as a Stacks address with a specific "smart contract" version byte. The
+smart contract identifier is a _principal_.
+
+For example, to call the function `register-name` in a smart contract,
+you would use:
+
+```scheme
+(contract-call
+    'SC3H92H297DX3YDPFHZGH90G8Z4NPH4VE8E83YWAQ
+    'register-name
+    name-to-register)
+```
+
+This function returns a boolean-- the return value of the called smart
+contract function. Note that if a called smart contract returns
+`false`, it is guaranteed to not alter any smart contract state
+whatsoever. Of course, any transaction fees paid for the execution
+of that function will not be returned.
```

Comment:
But, this is true in general, right?  Once the network agrees to process your transaction, your fee is effectively consumed the minute the code starts executing.

Reply:
Yep, just wanted to highlight that point here as well.